src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -28,45 +140,79 @@ public Future<TokenUsage> getTokenStats(ProxyContext context) {
         return traceContext.getStats(context);
     }
 
-    public void endSpan(ProxyContext context) {
+    public Future<Void> endSpan(ProxyContext context) {
         ApiKeyData apiKeyData = context.getApiKeyData();
         if (apiKeyData.getPerRequestKey() == null) {
             traceIdToContext.remove(context.getTraceId());
         } else {
             TraceContext traceContext = traceIdToContext.get(context.getTraceId());
             if (traceContext != null) {
-                traceContext.endSpan(context);
+                return traceContext.endSpan(context);
             }
         }
+        return Future.succeededFuture();
     }
 
     @VisibleForTesting
     Map<String, TraceContext> getTraces() {
         return traceIdToContext;
     }
 
-    private static class TraceContext {
+    @VisibleForTesting
+    Map<String, TrackerData> getRegisteredTrackers() {
+        return registeredTrackers;
+    }
+
+    private class TraceContext {
         Map<String, TokenStats> spans = new HashMap<>();
 
-        synchronized void addSpan(ProxyContext context) {
+        synchronized Future<Void> addSpan(ProxyContext context) {
             String spanId = context.getSpanId();
             String parentSpanId = context.getParentSpanId();
-            TokenStats tokenStats = new TokenStats(new TokenUsage(), Promise.promise(), new ArrayList<>());
+            TokenStats tokenStats = new TokenStats(new TokenUsage(), new HashMap<>());
             spans.put(spanId, tokenStats);
             if (parentSpanId != null) {
-                TokenStats parent = spans.get(parentSpanId);
-                if (parent != null) {
-                    parent.children.add(tokenStats.promise.future());
-                }
+                StartSpanEvent event = new StartSpanEvent(trackerId, context.getTraceId(), parentSpanId, context.getSpanId());
+                return publish(Event.START_SPAN, event);
             }
+            return Future.succeededFuture();
+        }
+
+        synchronized Promise<TokenUsage> updateSpan(StartSpanEvent event) {
+            TokenStats tokenStats = spans.get(event.parentSpanId);
+            if (tokenStats == null) {
+                return null;
+            }
+            Promise<TokenUsage> promise = Promise.promise();
+            tokenStats.children.put(event.spanId, promise);
+            return promise;
+        }
+
+        synchronized Promise<TokenUsage> updateSpan(EndSpanEvent event) {
```

Comment:
Span could end with error (for some reason even though we get successful event?), how do we know when to end with error?

Reply:
In case of error the core send an empty tokenUsage: e.g. bug in the cost calculator. So the code treats all results as successful. This is how it works now. We always complete a promise successfully.